### DISCUSSION TITLE
Deterministic JS output and internal object representation

### DISCUSSION THREAD

**Discussion by MarkMcCulloh** [winglang/wing#887] [General]
1. Given a wing program, should the intermediate JS output always be the same if there are no changes to the execution environment?
2. Given a wing program, should the final output (terraform, for now) always be the same if there are no changes to the execution environment?

I think most would agree on (2), but I suspect there is some disagreement on (1). If you agree with (2) but not (1), how can we guarantee (2) without (1)?

This thought came up as a result of [usage of BTreeMap instead of HashMap](https://github.com/winglang/wing/blob/0eab316c6e32ef087fc1fe2972de85ad57a4e313/libs/wingc/src/capture.rs#L24) in wingc as an easy way to get deterministly ordered stuff, because otherwise the order may change during each wingc invocation. I think this should be the default way to store things that we plan on iterating for JSification and we should rely on it.

I don't see this solely as a way to ensure snapshot tests pass, but rather as a nice way to guarantee both (1) and (2). Are there objections to this, and if so what are some alternatives?

**Comment by ekeren** [comment]
Because we rely on terraform for deployment (currently), it is a preferred DX (including operators experience) not see changes in areas that we haven't really changed when a running `terraform plan`.

I would even claim that it is preferred DX to keep the format stable between all minor versions of wing compiler and sdk resources, and not just between two different executions with the same toolchain

Because lambda functions is part of the terraform configuration, I would say that `inflight` code is a subgroup of 2, which leaves very little incentive not to declare deterministic output

Important to note, that until we release a stable version, we are not backward compatible so we can continue to break both 1 and 2, but I agree it makes sense to build the toolchain that generates a deterministic output

**Comment by Chriscbr** [comment]
What are the benefits of having different intermediate JS output on each compile? I feel like consistent JS output (1) may be a prereq for consistent terraform output (2) in many situations.

**Comment by eladb** [comment]
No objections at all. -- I think deterministic both is important and Good.

**Comment by staycoolcall911** [comment]
We're all for deterministic output, of course. This discussion is mainly about whether this means we will change the underlying implementation of Wing's Map type to be an ordered Map impl type. We decided it makes sense to do it and made that change in [this PR](https://github.com/winglang/wing/pull/908).
